### ISSUE TITLE
posible memory leak

### ISSUE THREAD

**Issue by visox** [jhy/jsoup#1130]
Hi i am investigating my memory usage because after a while it becomes unbearable.

The project is deployed on a debian server with JDK 9 and 256GB ram, eventually it gets all the ram and crashes.
Its in scala.
The project is heavily parallel, for that i am using akka-streams and scala-cats(io monad)

A did also run the project on my mac so that i can play play with it, eventually i also tried jvisualVM

<img width="982" alt="screen shot 2018-10-06 at 2 29 26 pm" src="https://user-images.githubusercontent.com/8051731/46571392-98198a00-c974-11e8-91f4-0252a655091a.png">

on the mac i also have JDK9

So i noticed jsoup on the top (besides primitives) and wondering if its normal.

(What would be to normal memory conpsumtion for one document ?)

i was using the latest jsoup (custom build) from 28.12.2017 after i read this https://github.com/jhy/jsoup/issues/938

Today i replaced it with the latest version from maven, will see if that helps (but probably wont)

the code
```
    val result = IO {

      val con = Jsoup.connect(encodedUrl)

      con.headers(headers.asJava)
      agent.foreach(con.userAgent)
      con.followRedirects(true)
      con.validateTLSCertificates(false)
      con.ignoreHttpErrors(true)

      con.maxBodySize(1024 * 1024 * 3)
      con.cookies(lastCookies.asJava)
      con.referrer(referrer)

      setup.map(_.proxy.toProxy).foreach(con.proxy)
      con.timeout(connectionTimeout.toMillis.toInt)

      val r = con.execute()

      lastCookies = r.cookies().asScala.toMap[String, String]

      val parsed = r.parse()
      
      ParsedDocument(parsed, url)
    }.unsafeTimed(connectionTimeout.plus(1 second))(executionContext)
```

i also sometimes create dummy documents `Document.createShell(url)`

I am handling a variety of exceptions, some lead to retries...

one way i run it (if its important) is `nohup nice -n 19 java -Xmx220g -jar dataCollect.jar 100 &`

Is there something that needs to be closed or released maybe ?

The problem could be somewhere else but i am desperate,
help me obi wan you're my only hope

**Comment by jhy** [comment]
Jsoup doesn't maintain anything outside of a document's GC lifecycle. Have benchmarked and profiled extensively and I'm quite confident there are no leaks. The issue you point to is not a leak either (and I'm not sure why you need a custom build when the memory use improvement is released in 1.11.1).

My guess would be that you're holding Documents or other DOM elements somewhere in your code.

I would suggest using a Java profiler to look at what is retaining nodes. I find YourKit Profiler very helpful.

**Comment by visox** [comment]
k thx
